LibGit2: add resolve_url to RemoteCallbacksStruct for LibGit2 0.99.0

diabonas opened the pull request:
Upstream commit https://github.com/libgit2/libgit2/commit/59647e1ad095f80112918971b7bbe05adfaf8c3c ("remote: add callback to resolve URLs before connecting") introduced a new callback `resolve_url` in LibGit2 0.99.0. Even though it is not currently used in Julia, it needs to be accounted for to get the correct size for `FetchOptionsStruct`. An incorrectly aligned `FetchOptionsStruct` leads to error messages like "invalid version 0 on git_proxy_options" when trying to use the latest LibGit2 version.

Fixes: GH-35043
Fixes: [FS#65540](https://bugs.archlinux.org/task/65540)

nalimilan committed:
LibGit2: amend GitError enum

Since upstream commit
https://github.com/libgit2/libgit2/commit/e9cef7c4b16b2cb572ac19fcd39217f7934cfa99
("http: introduce GIT_ERROR_HTTP") an invalid content type yields a
GIT_ERROR_HTTP instead of a GIT_ERROR_NET error. Update the enum to
include this new error so that the unit test for LibGit2 doesn't fail
with "invalid value for Enum Class: 34".

diabonas committed:
LibGit2: add resolve_url to RemoteCallbacksStruct for LibGit2 0.99.0

Upstream commit
https://github.com/libgit2/libgit2/commit/59647e1ad095f80112918971b7bbe05adfaf8c3c
("remote: add callback to resolve URLs before connecting") introduced a
new callback "resolve_url" in LibGit2 0.99.0. Even though it is not
currently used in Julia, it needs to be accounted for to get the correct
size for FetchOptionsStruct. An incorrectly aligned FetchOptionsStruct
leads to error messages like "invalid version 0 on git_proxy_options"
when trying to use the latest LibGit2 version.

nalimilan commented:
Ah too bad, looks like we were both debugging this at the same time. Though this isn't enough to get tests to pass on 0.99.0, see https://github.com/JuliaLang/julia/pull/35233.

diabonas commented:
Interesting, I ran the test suite while building the Arch Linux package, but apparently the [LibGit2 tests are disabled there](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/julia&id=c1675e6bfa179a54e2c00ce189997dfb91d0ae84#n76)... The only reason why I could see the tests failing is the changed error message https://github.com/JuliaLang/julia/pull/35233/commits/fa4481b0ef8ff668da71d64ea41cc8d209775bee#diff-d9d05c74120373a1756ee4a7e86c7cfeR3046. The other changes in your PR seem to be implemented in a backward-compatible way in LibGit2 0.99.0, so are not strictly required (but certainly make sense for the latest version).

That said, your PR is a lot more comprehensive than mine, so seems to be the better option going forward, if backwards-compatibility with older LibGit2 versions is not required (I don't know how this is usually handled in Julia).

diabonas committed:
LibGit2: update error message checking for LibGit2 0.99.0

Upstream commit
https://github.com/libgit2/libgit2/commit/b9c5b15a7958ab4ecb83504a6d858efd18610297
("http: use the new httpclient") changed the error message to include
additional quotes. Relax the unit test to allow these if present.

diabonas commented:
For completeness I included the two changes from your PR that are necessary to make the unit test pass in order to make this a minimal self-contained patch that works with all versions of LibGit2.

diabonas commented:
The [test failure](https://build.julialang.org/#/builders/65/builds/9022) for the Windows build appears to be unrelated:
```
Error in testset errorshow:
Test Failed at D:\buildbot\worker-tabularasa\tester_win32\build\share\julia\test\errorshow.jl:640
  Expression: occursin("the last 2 lines are repeated 5000 more times", output[5])
   Evaluated: occursin("the last 2 lines are repeated 5000 more times", " ... (the last 2 lines are repeated 4998 more times)")
Error in testset errorshow:
Test Failed at D:\buildbot\worker-tabularasa\tester_win32\build\share\julia\test\errorshow.jl:641
  Expression: (output[6])[1:8] == " [10003]"
   Evaluated: " [9999] " == " [10003]"
```

nalimilan reviewed (approved):
Thanks! I guess we should backport this to 1.4, then we can sort out the libgit2 0.99.1 build issues in #35233 for master.
